[docs] Add a brief explanation about linters (#12946)
conan-io/conan-center-index: docs/linters.md

## Conan Center Index Linters

Some linter configuration files are available in the folder [linter](../linter), which are executed by Github Actions to improve recipe quality.
They consume python scripts which are executed to fit CCI rules. Those scripts use [astroid](https://github.com/PyCQA/astroid) and
[pylint](https://pylint.pycqa.org/en/latest/) classes to parse Conan recipe files and manage their warnings and errors.

The pylint by itselt is not able to find Conan Center Index rules, so astroid is used to iterate over conanfiles content and
validate CCI conditions. Also, pylint uses [rcfile](https://pylint.pycqa.org/en/latest/user_guide/configuration/index.html)
(a configuration file based on [toml](https://toml.io/en/) format) to configure plugins, warnings and errors which should be enabled or disabled.

Also, the Github [code review](https://github.com/features/code-review) is integrated with the pylint output,
parsed by [recipe_linter.json](../linter/recipe_linter.json), then presented to all users on the tab `Files changed`.

> Of course, if wanto to run any linter locally, before pushing your code, read [Running the linter locally](v2_linter.md#running-the-linter-locally).

Reviewer: ```suggestion
### Running Linters

Of course, you should run any linter locally, before pushing your code, read [Running the linter locally](v2_linter.md#running-the-linter-locally) for instructions.
```

validate CCI conditions. Also, pylint uses [rcfile](https://pylint.pycqa.org/en/latest/user_guide/configuration/index.html)
(a configuration file based on [toml](https://toml.io/en/) format) to configure plugins, warnings and errors which should be enabled or disabled.

Also, the Github [code review](https://github.com/features/code-review) is integrated with the pylint output,
parsed by [recipe_linter.json](../linter/recipe_linter.json), then presented to all users on the tab `Files changed`.

### Running Linters

Of course, you should run any linter locally, before pushing your code, read [Running the linter locally](v2_linter.md#running-the-linter-locally) for instructions.

### Pylint configuration files

#### [Pylint Recipe](../linter/pylintrc_recipe)

This rcfile lists plugins and rules to be executed over all recipes (not test package) and validate them.

#### [Pylint Test Package Recipe](../linter/pylintrc_testpackage)

This rcfile lists plugins and rules to be executed over all recipes in test package folders only:

### Linter Warning and Errors

Here is the list of current warning and errors provided by pylint, when using CCI configuration.

> #### W9006 - conan-import-conanfile: ConanFile should be imported from conan

Author: ```suggestion
#### E9006 - conan-import-conanfile: ConanFile should be imported from conan
```
https://github.com/conan-io/conan-center-index/commit/c3f7f5af0aaa957d287e9befd7f1028bac586a0a

## Conan Center Index Linters

Some linter configuration files are available in the folder [linter](../linter), which are executed by Github Actions to improve recipe quality.

> They consume python scripts which are executed to fit CCI rules. Those scripts use `astroid` and `pylint` classes.

Reviewer 2: links to astroid/pylint would be nice

Author: Done.

To run any linter locally, read [Running the linter locally](v2_linter.md#running-the-linter-locally).

#### [Pylint Recipe](../linter/pylintrc_recipe)

> This rcfile lists plugins and rules to be executed over all recipes (not test package) and validate them:

Reviewer 2: what is rcfile?

Author: Done.
